File: python/grep/grep.py
from collections import namedtuple
import re

def grep(pattern, flags, files):
    grep_lines = []
    for f in files:
        grep_lines += process_single_file(f, pattern, flags)
    return format_results(grep_lines, flags, len(files) > 1)

def process_single_file(file, pattern, flags):
    grep_lines = []
    regex = re.compile(f'.*{pattern}')
    if '-i' in flags:
        regex = re.compile(f'.*{pattern}', flags=re.IGNORECASE)
    
    f = open(file, 'r')
    counter = 1
    for line in f.readlines():
        matched = False
        if '-x' in flags:
            if line.rstrip().lower() == pattern.lower():
                matched = True
        elif regex.match(line):
            matched = True
        
        if '-v' in flags:
            if not matched:
                grep_lines.append(MatchLine(counter, line, file))
        elif matched:
            grep_lines.append(MatchLine(counter, line, file))
        
        counter += 1
    
    return grep_lines

def format_results(lines, flags, include_file_name):
    if '-l' in flags:
        files = []
        for l in lines:
            new_file = l.file_name+'\n'
            if new_file not in files:
                files.append(new_file)

        return ''.join(files)
    elif '-n' in flags:
        if include_file_name:
            return ''.join(['%s:%s:%s' % (l.file_name, l.line_number, l.line_text) for l in lines])
        else:
            return ''.join(['%s:%s' % (l.line_number, l.line_text) for l in lines])
    else:
        if include_file_name:
            return ''.join(['%s:%s' % (l.file_name, l.line_text) for l in lines])
        else:
            return ''.join([l.line_text for l in lines])

MatchLine = namedtuple('MatchLine', 'line_number, line_text, file_name')

File: ruby/grep/grep.rb
class Grep
  def self.grep(pattern, flags, files)
    grep_lines = []
    files.each { |f| grep_lines += process_single_file(f, pattern, flags) }
    format_results(grep_lines, flags, files.size > 1)
  end

  def self.process_single_file(file, pattern, flags)
    grep_lines = []
    regexp_flags = flags.include?('-i') ? Regexp::IGNORECASE : nil
    reg = Regexp.new(pattern, regexp_flags)

    File.open(file).each.with_index do |l, i|
      matched = false
      if flags.include?('-x')
        matched =  true if l.rstrip.downcase == pattern.downcase
      elsif reg.match(l)
        matched =  true
      end

      if flags.include?('-v')
        grep_lines << MatchLine.new(i + 1, l.rstrip, file) unless matched
      elsif matched
        grep_lines << MatchLine.new(i + 1, l.rstrip, file)
      end
    end

    grep_lines
  end

  def self.format_results(grep_lines, flags, include_file_name)
    format_results = if flags.include?('-l')
                       grep_lines.map(&:file_name).uniq
                     elsif flags.include?('-n')
                       if include_file_name
                         grep_lines.map { |e| '%s:%s:%s' % [e.file_name, e.line_number, e.line_text] }
                       else
                         grep_lines.map { |e| '%s:%s' % [e.line_number, e.line_text] }
                       end
                     else
                       grep_lines.map { |e| include_file_name ? '%s:%s' % [e.file_name, e.line_text] : e.line_text }
                     end

    format_results.join("\n")
  end

  class MatchLine
    attr_reader :line_number, :line_text, :file_name

    def initialize(line_number, line_text, file_name)
      @line_number = line_number
      @line_text = line_text
      @file_name = file_name
    end
  end
end
